services/wallet/balance/balance_cache.go

```diff
@@ -41,253 +35,187 @@ type CacheIface interface {
 }
 
-func (b *Cache) AddBalance(account common.Address, chainID uint64, blockNumber *big.Int, balance *big.Int) {
+func (b *genericCache[B, _, _]) AddBalance(account common.Address, chainID uint64, blockNumber *big.Int, balance *big.Int) {
 	b.rw.Lock()
 	defer b.rw.Unlock()
 
 	_, exists := b.balances[account]
 	if !exists {
-		b.balances[account] = make(map[uint64]map[uint64]*big.Int)
+		b.balances[account] = make(map[uint64]B)
 	}
 
 	_, exists = b.balances[account][chainID]
 	if !exists {
-		b.balances[account][chainID] = make(map[uint64]*big.Int)
-	}
-
-	b.balances[account][chainID][blockNumber.Uint64()] = balance
-}
-
-func (b *Cache) BalanceAt(ctx context.Context, client Reader, account common.Address, blockNumber *big.Int) (*big.Int, error) {
-	cachedBalance := b.GetBalance(account, client.NetworkID(), blockNumber)
-	if cachedBalance != nil {
-		return cachedBalance, nil
-	}
-	balance, err := client.BalanceAt(ctx, account, blockNumber)
-	if err != nil {
-		return nil, err
+		b.balances[account][chainID] = reflect.New(reflect.TypeOf(b.balances[account][chainID]).Elem()).Interface().(B)
+		b.balances[account][chainID].init()
 	}
-	b.AddBalance(account, client.NetworkID(), blockNumber, balance)
 
-	return balance, nil
+	b.balances[account][chainID].set(blockNumber.Uint64(), balance)
 }
 
-func (b *Cache) GetNonce(account common.Address, chainID uint64, blockNumber *big.Int) *int64 {
+func (b *genericCache[_, N, NR]) GetNonce(account common.Address, chainID uint64, blockNumber *big.Int) *int64 {
 	b.rw.RLock()
 	defer b.rw.RUnlock()
 
-	if b.nonces[account] == nil || b.nonces[account][chainID] == nil {
+	_, exists := b.nonces[account]
+	if !exists {
 		return nil
 	}
-	return b.nonces[account][chainID][blockNumber.Uint64()]
-}
 
-func (b *Cache) Cache() CacheIface {
-	return b
-}
+	_, exists = b.nonces[account][chainID]
+	if !exists {
+		return nil
+	}
 
-func (b *Cache) sortRanges(account common.Address, chainID uint64) {
-	keys := make([]int, 0, len(b.nonceRanges[account][chainID]))
-	for k := range b.nonceRanges[account][chainID] {
-		keys = append(keys, int(k))
+	nonce := b.nonces[account][chainID].get(blockNumber.Uint64())
+	if nonce != nil {
+		return nonce
 	}
 
-	sort.Ints(keys) // This will not work for keys > 2^31
+	return b.findNonceInRange(account, chainID, blockNumber)
+}
 
-	ranges := []nonceRange{}
-	for _, k := range keys {
-		r := b.nonceRanges[account][chainID][int64(k)]
-		ranges = append(ranges, r)
+func (b *genericCache[_, N, NR]) AddNonce(account common.Address, chainID uint64, blockNumber *big.Int, nonce *int64) {
+	b.rw.Lock()
+	defer b.rw.Unlock()
+
+	_, exists := b.nonces[account]
+	if !exists {
+		b.nonces[account] = make(map[uint64]N)
 	}
 
-	_, exists := b.sortedRanges[account]
+	_, exists = b.nonces[account][chainID]
 	if !exists {
-		b.sortedRanges[account] = make(map[uint64][]nonceRange)
+		b.nonces[account][chainID] = reflect.New(reflect.TypeOf(b.nonces[account][chainID]).Elem()).Interface().(N)
```

Comment:
I have no idea how common it is to use reflection in such cases in Go. Still I have some discomfort while seeing it and would prefer having simple builder functions provided by `genericCache` implementation instead. 

It might be an idiomatic way to do things here (again idk what are performance implications), so if it is fine just ignore the comment.

Reply:
provided we have simple builders for all three cache instances we could probably avoid parametrisation of `genericCache`. That also might make it a bit more straightforward not only in terms of getting rid of some reflection code but just general readability

Reply:
we have 
```go
type genericCache[B cacheIface[uint64, *big.Int], N cacheIface[uint64, *int64], NR cacheIface[int64, nonceRange]] struct {
	nonceRangeCache[NR]

	// balances maps an address and chain to a cache of a block number and the balance of this particular address on the chain
	balances addressChainMap[B]
	nonces   addressChainMap[N]
	rw       sync.RWMutex
}
```

which could be something like this for removing parameters from `genericCache` and getting rid of reflection. It is not going to be less verbose, just arguably a bit more straightforward. 

```go
type cacheBuilderInterface {
    newBalance() cacheIface[uint64, *big.Int]
    newNonce() cacheIface[uint64, *int64]
    newNonceRange() cacheIface[int64, nonceRange]
}

type genericCache struct {
	nonceRangeCache[cacheIface[int64, nonceRange]]

	// balances maps an address and chain to a cache of a block number and the balance of this particular address on the chain
	balances addressChainMap[cacheIface[uint64, *big.Int]]
	nonces   addressChainMap[cacheIface[uint64, *int64]]
        builder cacheBulderInterface   
	rw       sync.RWMutex
}
```

Reply:
```go
//then
b.nonces[account][chainID] = reflect.New(reflect.TypeOf(b.nonces[account][chainID]).Elem()).Interface().(N)
//becomes
b.nonces[account][chainID] = b.builder.NewNonce()
```